test/recipes/90-test_store.t

```diff
@@ -44,21 +44,32 @@ my @generated_files =
      "rsa-key-pkcs8-pbes2-sha1.pem", "rsa-key-pkcs8-pbes2-sha1.der",
      "rsa-key-sha1-3des-sha1.p12", "rsa-key-sha1-3des-sha256.p12",
      "rsa-key-aes256-cbc-sha256.p12",
-     "rsa-key-md5-des-sha1.p12",
-     "rsa-key-aes256-cbc-md5-des-sha256.p12",
      "rsa-key-pkcs8-pbes2-sha256.pem", "rsa-key-pkcs8-pbes2-sha256.der",
-     "rsa-key-pkcs8-pbes1-md5-des.pem", "rsa-key-pkcs8-pbes1-md5-des.der",
+    );
+my $use_md5 = 0; # does not work: !disabled("md5");
+push(@generated_files, (
+    "rsa-key-md5-des-sha1.p12",
+    "rsa-key-aes256-cbc-md5-des-sha256.p12",
+    "rsa-key-pkcs8-pbes1-md5-des.pem", "rsa-key-pkcs8-pbes1-md5-des.der"
+     )) if $use_md5;
+my $use_dsa = !disabled("dsa")
+              && run(app(["openssl", "dsaparam", "-genkey", "1024"]));
```

Comment:
Why is running dsaparam here necessary?

Reply:
This is because I found that even when DSA is enabled the key generation may fail
(maybe because 1024 is not any more considered secure enough).

Reply:
Is there a better way to test this at this point?

Reply:
Again my question: how to check with Perl if DSA-1024 is (not) deprecated?

Reply:
What if DSA-2048 was used instead?

Reply:
> What if DSA-2048 was used instead?

Good thought. I've just tried it out but this did not help.

Yet since fixing the applicability of those test is just a side issue that I tried to tackle along with fixing the mem leaks in `try_decode_PKCS12()` (and the docs) I suggest leaving this topic for now, also because it's not urgent and there is already an approval.

Reply:
All this discussion doesn't explain why you have that run call here.  Isn't the one in the `init` function enough?

Reply:
> > When loading PKCS#12 files that include a cert/ca chain the chain container was not deallocated.
> 
> Looking back, I'm surprised we haven't seen that already. Is there a way to demonstrate this with existing master, or even 1.1.1?

With the current master it's hard to reproduce because my commit 3c38fa4b797848a76b83f91e423de470adbb3b61, which is already merged, includes a preliminary fix, while this PR also fixes the mem leaks on malloc failures.

Moreover, the pkcs12 app does not use the STORE API for loading PKCS#12 files, and in 1.1.1 apparently the only app using it is `storeutl`.
And so far the tests did not include a chain in the PKCS#12 file(s) parsed.

Here is a way to demonstrate the mem leak for the 1.1.1 master if mdebug is enabled:
```
apps/openssl pkcs12 -inkey test/certs/interCA.key -in test/certs/interCA.pem -chain -CAfile test/certs/rootCA.pem -export -descert --passout pass: -out test.p12
apps/openssl storeutl test.p12
```

Reply:
> All this discussion doesn't explain why you have that run call here. Isn't the one in the `init` function enough?

The problem is that (at least meanwhile) the init function fails due to some algs not available and then the test 'succeeds' simply because the test cases are not run and the test setup is not well designed to fail in such a case:
```
HARNESS_VERBOSE_FAILURE=yes make -j4 test TESTS=test_store V=1

make depend && make _tests
make[1]: Entering directory '/home/david/Siemens/proj/PKI/openssl/prepare1'
make[1]: Leaving directory '/home/david/Siemens/proj/PKI/openssl/prepare1'
make[1]: Entering directory '/home/david/Siemens/proj/PKI/openssl/prepare1'
( SRCTOP=. \
  BLDTOP=. \
  PERL="/usr/bin/perl" \
  EXE_EXT= \
  /usr/bin/perl ./test/run_tests.pl test_store )
90-test_store.t .. 
# The results of this test will end up in test-runs/test_store
Error, DSA key generation failed
../../util/wrap.pl ../../apps/openssl dsaparam -genkey 1024 => 1
# -----BEGIN EC PARAMETERS-----
# BggqhkjOPQMBBw==
# -----END EC PARAMETERS-----
# -----BEGIN EC PRIVATE KEY-----
# MHcCAQEEIPjXwljI5SKHSjCu7lAKtqfzEBHg3aUlOtBNGDyQ0B8yoAoGCCqGSM49
# AwEHoUQDQgAECpWVBwFOV0OD2/aXIfJ7KhruuYA4LLYCVMnEQXv6w/lcNepWvyYj
# Ae4jy5j8llmlZLzbQkIrRTR9ILllAkOlRA==
# -----END EC PRIVATE KEY-----
../../util/wrap.pl ../../apps/openssl ecparam -genkey -name prime256v1 => 0
1..177
../../../util/wrap.pl ../../../apps/openssl genrsa -out rsa-key-pkcs1.pem 2432 => 0
writing RSA key
../../../util/wrap.pl ../../../apps/openssl rsa -passout 'pass:password' -aes128 -in rsa-key-pkcs1.pem -out rsa-key-pkcs1-aes128.pem => 0
../../../util/wrap.pl ../../../apps/openssl ecparam -genkey -name prime256v1 -out ec-key-pkcs1.pem => 0
read EC key
writing EC key
../../../util/wrap.pl ../../../apps/openssl ec -passout 'pass:password' -aes128 -in ec-key-pkcs1.pem -out ec-key-pkcs1-aes128.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -nocrypt -in rsa-key-pkcs1.pem -out rsa-key-pkcs8.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -nocrypt -in ec-key-pkcs1.pem -out ec-key-pkcs8.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -passout 'pass:password' -v1 pbeWithSHA1And3-KeyTripleDES-CBC -in rsa-key-pkcs8.pem -out rsa-key-pkcs8-pbes1-sha1-3des.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -passout 'pass:password' -v1 pbeWithSHA1And3-KeyTripleDES-CBC -in rsa-key-pkcs8.pem -out rsa-key-pkcs8-pbes1-md5-des.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -passout 'pass:password' -v2 aes256 -v2prf hmacWithSHA1 -in rsa-key-pkcs8.pem -out rsa-key-pkcs8-pbes2-sha1.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -passout 'pass:password' -v2 aes256 -v2prf hmacWithSHA1 -in ec-key-pkcs8.pem -out ec-key-pkcs8-pbes2-sha1.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs8 -topk8 -passout 'pass:password' -v2 aes256 -v2prf hmacWithSHA256 -in rsa-key-pkcs8.pem -out rsa-key-pkcs8-pbes2-sha256.pem => 0
Generating a RSA private key
.......................................................................................................++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
..............++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
writing new private key to 'cakey.pem'
-----
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Common Name []:CA
../../../util/wrap.pl ../../../apps/openssl req -x509 -config ../../../test/recipes/90-test_store_data/ca.cnf -nodes -out cacert.pem -keyout cakey.pem => 0
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Common Name []:A user
User ID []:test
../../../util/wrap.pl ../../../apps/openssl req -new -config ../../../test/recipes/90-test_store_data/user.cnf -key rsa-key-pkcs8.pem -out rsa-cert.csr => 0
Request self-signature ok
subject=CN = A user, UID = test
Getting CA Private Key
../../../util/wrap.pl ../../../apps/openssl x509 -days 3650 -CA cacert.pem -CAkey cakey.pem -set_serial 1591182361 -req -in rsa-cert.csr -out rsa-cert.pem => 0
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Common Name []:A user
User ID []:test
../../../util/wrap.pl ../../../apps/openssl req -new -config ../../../test/recipes/90-test_store_data/user.cnf -key ec-key-pkcs8.pem -out ec-cert.csr => 0
Request self-signature ok
subject=CN = A user, UID = test
Getting CA Private Key
../../../util/wrap.pl ../../../apps/openssl x509 -days 3650 -CA cacert.pem -CAkey cakey.pem -set_serial 1591182361 -req -in ec-cert.csr -out ec-cert.pem => 0
../../../util/wrap.pl ../../../apps/openssl pkcs12 -inkey rsa-key-pkcs8.pem -in rsa-cert.pem -passout 'pass:password' -chain -CAfile cacert.pem -export -macalg SHA1 -certpbe pbeWithSHA1And3-KeyTripleDES-CBC -keypbe pbeWithSHA1And3-KeyTripleDES-CBC -out rsa-key-sha1-3des-sha1.p12 => 0
../../../util/wrap.pl ../../../apps/openssl pkcs12 -inkey rsa-key-pkcs8.pem -in rsa-cert.pem -passout 'pass:password' -chain -CAfile cacert.pem -export -macalg SHA256 -certpbe pbeWithSHA1And3-KeyTripleDES-CBC -keypbe pbeWithSHA1And3-KeyTripleDES-CBC -out rsa-key-sha1-3des-sha256.p12 => 0
../../../util/wrap.pl ../../../apps/openssl pkcs12 -inkey rsa-key-pkcs8.pem -in rsa-cert.pem -passout 'pass:password' -chain -CAfile cacert.pem -export -macalg SHA256 -certpbe AES-256-CBC -keypbe AES-256-CBC -out rsa-key-aes256-cbc-sha256.p12 => 0
C087AAB8427F0000:error::digital envelope routines:(unknown function):fetch failed:crypto/evp/evp_fetch.c:307:Default library context, Algorithm (DES-CBC), Properties ()
C087AAB8427F0000:error::digital envelope routines:(unknown function):initialization error:crypto/evp/evp_enc.c:176:
C087AAB8427F0000:error::digital envelope routines:(unknown function):keygen failure:crypto/evp/evp_pbe.c:132:
C087AAB8427F0000:error::PKCS12 routines:(unknown function):pkcs12 algor cipherinit error:crypto/pkcs12/p12_decr.c:36:
C087AAB8427F0000:error::PKCS12 routines:(unknown function):encrypt error:crypto/pkcs12/p12_decr.c:133:
C087AAB8427F0000:error::PKCS12 routines:(unknown function):encrypt error:crypto/pkcs12/p12_add.c:119:
../../../util/wrap.pl ../../../apps/openssl pkcs12 -inkey rsa-key-pkcs8.pem -in rsa-cert.pem -passout 'pass:password' -chain -CAfile cacert.pem -export -macalg SHA1 -certpbe pbeWithMD5AndDES-CBC -keypbe pbeWithMD5AndDES-CBC -out rsa-key-md5-des-sha1.p12 => 1
ok 1 # skip failed initialisation
ok 2 # skip failed initialisation
ok 3 # skip failed initialisation
ok 4 # skip failed initialisation
ok 5 # skip failed initialisation
ok 6 # skip failed initialisation
ok 7 # skip failed initialisation
ok 8 # skip failed initialisation
ok 9 # skip failed initialisation
ok 10 # skip failed initialisation
ok 11 # skip failed initialisation
ok 12 # skip failed initialisation
ok 13 # skip failed initialisation
ok 14 # skip failed initialisation
ok 15 # skip failed initialisation
ok 16 # skip failed initialisation
ok 17 # skip failed initialisation
ok 18 # skip failed initialisation
ok 19 # skip failed initialisation
ok 20 # skip failed initialisation
ok 21 # skip failed initialisation
ok 22 # skip failed initialisation
ok 23 # skip failed initialisation
ok 24 # skip failed initialisation
ok 25 # skip failed initialisation
ok 26 # skip failed initialisation
ok 27 # skip failed initialisation
ok 28 # skip failed initialisation
ok 29 # skip failed initialisation
ok 30 # skip failed initialisation
ok 31 # skip failed initialisation
ok 32 # skip failed initialisation
ok 33 # skip failed initialisation
ok 34 # skip failed initialisation
ok 35 # skip failed initialisation
ok 36 # skip failed initialisation
ok 37 # skip failed initialisation
ok 38 # skip failed initialisation
ok 39 # skip failed initialisation
ok 40 # skip failed initialisation
ok 41 # skip failed initialisation
ok 42 # skip failed initialisation
ok 43 # skip failed initialisation
ok 44 # skip failed initialisation
ok 45 # skip failed initialisation
ok 46 # skip failed initialisation
ok 47 # skip failed initialisation
ok 48 # skip failed initialisation
ok 49 # skip failed initialisation
ok 50 # skip failed initialisation
ok 51 # skip failed initialisation
ok 52 # skip failed initialisation
ok 53 # skip failed initialisation
ok 54 # skip failed initialisation
ok 55 # skip failed initialisation
ok 56 # skip failed initialisation
ok 57 # skip failed initialisation
ok 58 # skip failed initialisation
ok 59 # skip failed initialisation
ok 60 # skip failed initialisation
ok 61 # skip failed initialisation
ok 62 # skip failed initialisation
ok 63 # skip failed initialisation
ok 64 # skip failed initialisation
ok 65 # skip failed initialisation
ok 66 # skip failed initialisation
ok 67 # skip failed initialisation
ok 68 # skip failed initialisation
ok 69 # skip failed initialisation
ok 70 # skip failed initialisation
ok 71 # skip failed initialisation
ok 72 # skip failed initialisation
ok 73 # skip failed initialisation
ok 74 # skip failed initialisation
ok 75 # skip failed initialisation
ok 76 # skip failed initialisation
ok 77 # skip failed initialisation
ok 78 # skip failed initialisation
ok 79 # skip failed initialisation
ok 80 # skip failed initialisation
ok 81 # skip failed initialisation
ok 82 # skip failed initialisation
ok 83 # skip failed initialisation
ok 84 # skip failed initialisation
ok 85 # skip failed initialisation
ok 86 # skip failed initialisation
ok 87 # skip failed initialisation
ok 88 # skip failed initialisation
ok 89 # skip failed initialisation
ok 90 # skip failed initialisation
ok 91 # skip failed initialisation
ok 92 # skip failed initialisation
ok 93 # skip failed initialisation
ok 94 # skip failed initialisation
ok 95 # skip failed initialisation
ok 96 # skip failed initialisation
ok 97 # skip failed initialisation
ok 98 # skip failed initialisation
ok 99 # skip failed initialisation
ok 100 # skip failed initialisation
ok 101 # skip failed initialisation
ok 102 # skip failed initialisation
ok 103 # skip failed initialisation
ok 104 # skip failed initialisation
ok 105 # skip failed initialisation
ok 106 # skip failed initialisation
ok 107 # skip failed initialisation
ok 108 # skip failed initialisation
ok 109 # skip failed initialisation
ok 110 # skip failed initialisation
ok 111 # skip failed initialisation
ok 112 # skip failed initialisation
ok 113 # skip failed initialisation
ok 114 # skip failed initialisation
ok 115 # skip failed initialisation
ok 116 # skip failed initialisation
ok 117 # skip failed initialisation
ok 118 # skip failed initialisation
ok 119 # skip failed initialisation
ok 120 # skip failed initialisation
ok 121 # skip failed initialisation
ok 122 # skip failed initialisation
ok 123 # skip failed initialisation
ok 124 # skip failed initialisation
ok 125 # skip failed initialisation
ok 126 # skip failed initialisation
ok 127 # skip failed initialisation
ok 128 # skip failed initialisation
ok 129 # skip failed initialisation
ok 130 # skip failed initialisation
ok 131 # skip failed initialisation
ok 132 # skip failed initialisation
ok 133 # skip failed initialisation
ok 134 # skip failed initialisation
ok 135 # skip failed initialisation
ok 136 # skip failed initialisation
ok 137 # skip failed initialisation
ok 138 # skip failed initialisation
ok 139 # skip failed initialisation
ok 140 # skip failed initialisation
ok 141 # skip failed initialisation
ok 142 # skip failed initialisation
ok 143 # skip failed initialisation
ok 144 # skip failed initialisation
ok 145 # skip failed initialisation
ok 146 # skip failed initialisation
ok 147 # skip failed initialisation
ok 148 # skip failed initialisation
ok 149 # skip failed initialisation
ok 150 # skip failed initialisation
ok 151 # skip failed initialisation
ok 152 # skip failed initialisation
ok 153 # skip failed initialisation
ok 154 # skip failed initialisation
ok 155 # skip failed initialisation
ok 156 # skip failed initialisation
ok 157 # skip failed initialisation
ok 158 # skip failed initialisation
ok 159 # skip failed initialisation
ok 160 # skip failed initialisation
ok 161 # skip failed initialisation
ok 162 # skip failed initialisation
ok 163 # skip failed initialisation
ok 164 # skip failed initialisation
ok 165 # skip failed initialisation
ok 166 # skip failed initialisation
ok 167 # skip failed initialisation
ok 168 # skip failed initialisation
ok 169 # skip failed initialisation
ok 170 # skip failed initialisation
ok 171 # skip failed initialisation
ok 172 # skip failed initialisation
ok 173 # skip failed initialisation
ok 174 # skip failed initialisation
ok 175 # skip failed initialisation
ok 176 # skip failed initialisation
ok 177 # skip failed initialisation
ok
All tests successful.
Files=1, Tests=177,  2 wallclock secs ( 0.03 usr  0.00 sys +  1.31 cusr  0.22 csys =  1.56 CPU)
Result: PASS
```
```

Reply:
> The problem is that (at least meanwhile) the init function fails due to some algs not available and then the test 'succeeds' simply because the test cases are not run and the test setup is not well designed to fail in such a case:

So there are two part in this:
1.  detecting that a feature is disabled.
2.  detecting that usage of a feature works.

The first part, I totally agree should affect what files to expect, i.e. variables such as `$use_dsa = !disabled("dsa")` is exactly correct.
The second part should remain in `init`.  If it doesn't work, then the variable `$use_dsa` hasn't been checked properly, or one particular command invocation has failed and needs to be corrected.

The specific failure to init with pbeWithMD5AndDES-CBC can, I think, be solved like this:

``` diff
diff --git a/test/recipes/90-test_store.t b/test/recipes/90-test_store.t
index 3e2e69f439..3a58a2fd94 100644
--- a/test/recipes/90-test_store.t
+++ b/test/recipes/90-test_store.t
@@ -17,6 +17,8 @@ setup($test_name);
 
 my $mingw = config('target') =~ m|^mingw|;
 
+my $use_des = !disabled("des");
+
 my @noexist_files =
     ( "test/blahdiblah.pem",
       "test/blahdibleh.der" );
@@ -336,7 +338,9 @@ sub init {
                           my %pbes =
                               (
                                "sha1-3des" => "pbeWithSHA1And3-KeyTripleDES-CBC",
-                               "md5-des" => "pbeWithMD5AndDES-CBC",
+                               ($use_des
+                                ? ( "md5-des" => "pbeWithMD5AndDES-CBC" )
+                                : () ),
                                "aes256-cbc" => "AES-256-CBC",
                               );
                           my %macalgs =
```

Reply:
A trick to know with perl is that lists are flattened, i.e. `(1, (2, 3), 4)` becomes `(1, 2, 3, 4)`, and that `=> ` is syntax sugar for a list of a pair, i.e. `a => 1` is the same as `('a', 1)`

Reply:
Concerning the `openssl dsaparam` call used in the test init, I meanwhile found out that the key size of 1024 is no longer acceptable by the default and after changing this to an acceptable value such as 2048, dsaparam crashes. See the new bug report #12047.

As a workaround I resorted to using an already existing param file, 
which I found in `test/recipes/04-test_pem_data/dsaparam.pem`.

Reply:
Meanwhile I realized that the tests anyway should better not generate fresh RSA and DSA keys since this is just a waste of time/computations and the focus of these test is just reading various credential files via the OSSL_STORE API.
So I've changed this to using pre-computed RSA keys and pre-computed DSA parameters.